# Data Analysis with python
Automobile Data Set

## Data Wrangling

# Table of content


Identify and handle missing values 
Identify missing values
Deal with missing values
Correct data format
Data standardization
Data Normalization (centering/scaling)
Binning
Indicator variable

In [1]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns 
%matplotlib inline

In [2]:
filename="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DA0101EN/auto.csv"

In [3]:
columns_name= ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

In [4]:
df=pd.read_csv(filename,names=columns_name)

In [5]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

# How to work with missing data?

Steps for working with missing data:

dentify missing data

deal with missing data

correct data format

In [6]:
#conver ? into NAN
df.replace("?",np.NAN,inplace=True)
df.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

## Evaluating for Missing Data

to detect missing value there are two methods

1) .isnull()

2) .notnull()

In [7]:
missing_data=df.isnull()
missing_data.head()

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

## True stands for missing value

## count missing value

In [8]:
for column in missing_data.columns.values.tolist():
    print(column)
    print(missing_data[column].value_counts())

symboling
False    205
Name: symboling, dtype: int64
normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64
make
False    205
Name: make, dtype: int64
fuel-type
False    205
Name: fuel-type, dtype: int64
aspiration
False    205
Name: aspiration, dtype: int64
num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64
body-style
False    205
Name: body-style, dtype: int64
drive-wheels
False    205
Name: drive-wheels, dtype: int64
engine-location
False    205
Name: engine-location, dtype: int64
wheel-base
False    205
Name: wheel-base, dtype: int64
length
False    205
Name: length, dtype: int64
width
False    205
Name: width, dtype: int64
height
False    205
Name: height, dtype: int64
curb-weight
False    205
Name: curb-weight, dtype: int64
engine-type
False    205
Name: engine-type, dtype: int64
num-of-cylinders
False    205
Name: num-of-cylinders, dtype: int64
engine-size
False    205
Name: engine-size, dtype: int64
fuel-system
False    205
Name: 

## form the above information each column has 205 rows 

only 7 columns has the missing values

normalized-losses": 41 missing values

"num-of-doors": 2 missing values

"bore": 4 missing values

"stroke" : 4 missing values

"horsepower": 2 missing values

"peak-rpm": 2 missing values

"price": 4 missing values


# dealing with missing data

# Two methods to deal with
## 1) drop the row or column. 2) replace data with mean,frequency or other function



replace by mean is given as:

"normalized_losses"

"bore"

"stroke"

"horse-power"

"peak-rpm"



In [9]:
#first find mean of normalize losses
mean_norm_loss=df['normalized-losses'].astype("float").mean(axis=0)#axis 0 indecate row
print("mean of normalized-losses is :",mean_norm_loss)

mean of normalized-losses is : 122.0


Replace mean of normalized-losses with nan

In [10]:
df["normalized-losses"].replace(np.nan,mean_norm_loss,inplace=True)

In [11]:
mean_of_bore=df["bore"].astype("float").mean(axis=0)
print("mean of bore",mean_of_bore)

mean of bore 3.3297512437810957


In [12]:
df["bore"].replace(np.nan,mean_of_bore,inplace=True)

In [13]:
mean_of_stroke=df["stroke"].astype("float").mean(axis=0)
print("mean_of_stroke",mean_of_stroke)

mean_of_stroke 3.2554228855721337


In [14]:
df["stroke"].replace(np.nan,mean_of_stroke,inplace=True)


In [15]:
mean_horsepower=df["horsepower"].astype("float").mean(axis=0)
print("mean of horsepower",mean_horsepower)

mean of horsepower 104.25615763546799


In [16]:
df["horsepower"].replace(np.nan,mean_horsepower,inplace=True)

In [17]:
mean_of_peak_rpm=df["peak-rpm"].astype("float").mean(axis=0)

In [18]:
df["peak-rpm"].replace(np.nan,mean_of_peak_rpm,inplace=True)

# Replacement by frequency

first find the number of values in each colums
through "value.counts()" method

In [19]:
df["num-of-doors"].value_counts()

four    114
two      89
Name: num-of-doors, dtype: int64

# for the most number of value can be find as

In [20]:
df["num-of-doors"].value_counts().idxmax()

'four'

In [21]:
#now to replace with frquency through similer method as mean
df["num-of-doors"].replace(np.nan,'four',inplace=True)

In [22]:
df["num-of-doors"].head()

0     two
1     two
2     two
3    four
4    four
Name: num-of-doors, dtype: object

# drop the price which has nan value

In [23]:
#drop the row with nan value
df.dropna(subset=['price'],axis=0,inplace=True)
df.reset_index(drop=True,inplace=True)

In [24]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               122  alfa-romero       gas        std          two   
1          3               122  alfa-romero       gas        std          two   
2          1               122  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

# now we have data set with no missing values

# last step for data cleaning is to check the format

i.e the data is in (float int object etc) format

In [25]:
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

# as we can see above, some columns are not of the correct data type. Numerical variables should have type 'float' or 'int', and variables with strings such as categories should have type 'object'

In [26]:
df[["bore","stroke"]]=df[["bore","stroke"]].astype("float")
df["normalized-losses"]=df[["normalized-losses"]].astype("int")
df[["price"]]=df[["price"]].astype("float")
df[["peak-rpm"]]=df[["peak-rpm"]].astype("float")

In [27]:
df.dtypes

symboling              int64
normalized-losses      int32
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                 float64
stroke               float64
compression-ratio    float64
horsepower            object
peak-rpm             float64
city-mpg               int64
highway-mpg            int64
price                float64
dtype: object

now we clean our data

so our data has no missing value and has the right format